File: ScaleformUI_Lua/src/sidepanels/UIMissionDetailsPanel.lua
UIMissionDetailsPanel = setmetatable({}, UIMissionDetailsPanel)
UIMissionDetailsPanel.__index = UIMissionDetailsPanel
UIMissionDetailsPanel.__call = function() return "UIMissionDetailsPanel", "UIMissionDetailsPanel" end

function UIMissionDetailsPanel.New(side, title, color, inside, txd, txn)
    local _titleType, _titleColor
	if inside == -1 then
        _titleType = 1
    elseif inside then
        _titleType = 2
    else
        _titleType = 0
    end

    if color ~= -1 then
        _titleColor = color
    else
        _titleColor = Colours.NONE
    end

    _UIMissionDetailsPanel = {
        PanelSide = side,
        Title = title,
        TitleColor = _titleColor,
        TitleType = _titleType,
        TextureDict = txd or "",
        TextureName = txn or "",
        Items = {},
        ParentItem = nil
	}
	return setmetatable(_UIMissionDetailsPanel, UIMissionDetailsPanel)
end

function UIMissionDetailsPanel:SetParentItem(Item) -- required
	if Item() == "UIMenuItem" then
		self.ParentItem = Item
	else
		return self.ParentItem
	end
end

function UIMissionDetailsPanel:UpdatePanelTitle(title)
    self.Title = title

    if self.ParentItem ~= nil then
        local item = IndexOf(self.ParentItem.Base.ParentMenu.Items, self.ParentItem) - 1
        ScaleformUI.Scaleforms._ui:CallFunction("UPDATE_SIDE_PANEL_TITLE", false, item, title)
    end
end

function UIMissionDetailsPanel:UpdatePanelPicture(txd, txn)
    self.TextureDict = txd
    self.TextureName = txn

    if self.ParentItem ~= nil then
        local item = IndexOf(self.ParentItem.Base.ParentMenu.Items, self.ParentItem) - 1
        ScaleformUI.Scaleforms._ui:CallFunction("UPDATE_MISSION_DETAILS_PANEL_IMG", false, item, txd, txn)
    end
end

function UIMissionDetailsPanel:AddItem(newitem)
    table.insert(self.Items, newitem)
    if self.ParentItem ~= nil then
        local item = IndexOf(self.ParentItem.Base.ParentMenu.Items, self.ParentItem) - 1
        ScaleformUI.Scaleforms._ui:CallFunction("ADD_MISSION_DETAILS_DESC_ITEM", false, item, newitem.Type, newitem.TextLeft, newitem.TextRight, newitem.Icon, newitem.IconColor, newitem.Ticked)
    end
end

function UIMissionDetailsPanel:RemoveItemAt(index)	
    table.remove(self.Items, index)	
    if self.ParentItem ~= nil then	
        ScaleformUI.Scaleforms._ui:CallFunction("REMOVE_MISSION_DETAILS_DESC_ITEM", false, index - 1)	
    end	
end

File: ScaleformUI_V8/src/menus/UIMenu/sidepanels/DetailsPanel/uimissiondetailspanel.ts
import { UIMenuPanel } from "menus/UIMenu/panels/uimenupanel";
import { PanelSide } from "../uimenusidepanel";
import { SColor } from "elements/scolor";
import { UIFreemodeDetailsItem } from "./uifreemodedetailsitem";
import { ScaleformUI } from "scaleforms/scaleformui/main";
import { cache } from '@babel/traverse';

export class UIMissionDetailsPanel extends UIMenuPanel{

    private title:string;
    public PanelSide:PanelSide;
    public TitleColor:SColor;
    public TitleType:number;
    public TextureDict:string;
    public TextureName:string
    public Items:UIFreemodeDetailsItem[];

    public set Title(value:string){
        this.title = value;
        if (this.ParentItem != undefined && this.ParentItem.Parent != undefined && this.ParentItem.Parent.Visible)
        {
            ScaleformUI.Scaleforms._ui?.callFunction("UPDATE_SIDE_PANEL_TITLE", this.ParentItem.Parent.Pagination.GetScaleformIndex(this.ParentItem.Parent.Items.indexOf(this.ParentItem)), this.title);
        }
    }
    public get Title():string{
        return this.title;
    }

    public UpdatePanelPicture(txd:string, txn:string){
        this.TextureDict = txd;
        this.TextureName = txn;
        if (this.ParentItem != undefined && this.ParentItem.Parent != undefined && this.ParentItem.Parent.Visible)
        {
            let wid = this.ParentItem.Parent.Pagination.GetScaleformIndex(this.ParentItem.Parent.Items.indexOf(this.ParentItem));
            ScaleformUI.Scaleforms._ui?.callFunction("UPDATE_MISSION_DETAILS_PANEL_IMG", wid, this.TextureDict, this.TextureName);
        }
    }

    public AddItem(item:UIFreemodeDetailsItem){
        this.Items.push(item);
        if (this.ParentItem != undefined && this.ParentItem.Parent != undefined && this.ParentItem.Parent.Visible)
        {
            let wid = this.ParentItem.Parent.Pagination.GetScaleformIndex(this.ParentItem.Parent.Items.indexOf(this.ParentItem));
            ScaleformUI.Scaleforms._ui?.callFunction("ADD_MISSION_DETAILS_DESC_ITEM", wid, item.Type, item.TextLeft, item.TextRight, item.Icon, item.IconColor, item.Tick, item._labelFont.fontName, item._labelFont.fontId, item._rightLabelFont.fontName, item._rightLabelFont.fontId);
        }
    }

    public RemoveItem(idx:number){
        this.Items.splice(idx, 1);
        if (this.ParentItem != undefined && this.ParentItem.Parent != undefined && this.ParentItem.Parent.Visible)
        {
            let wid = this.ParentItem.Parent.Pagination.GetScaleformIndex(this.ParentItem.Parent.Items.indexOf(this.ParentItem));
            ScaleformUI.Scaleforms._ui?.callFunction("REMOVE_MISSION_DETAILS_DESC_ITEM", wid, idx);
        }

    }



    constructor(side:PanelSide, title:string, color:SColor, inside:number|boolean, txd:string, txn:string){
        super();
        let _titleType:number;
        let _titleColor:SColor;

        if (inside == -1)
            _titleType = 1
        else if (inside)
            _titleType = 2
        else
            _titleType = 0
        
    
        if (color != SColor.HUD_None)
            _titleColor = color
        else
            _titleColor = SColor.HUD_None
        
        this.PanelSide = side;
        this.Title = title;
        this.TitleColor = _titleColor;
        this.TitleType = _titleType;
        this.TextureDict = txd || "";
        this.TextureName = txn || "";
        this.Items = [];
    }


}